File: CharacterSheetCompanion/CharacterSheetCompanion/Services/SocketListeners/ActorListeners/ActorClassesListener.swift
//
//  ActorBaseDataListener.swift
//  CharacterSheetCompanion
//
//  Created by Zachary Sanders on 8/17/21.
//

import Combine
import Foundation
import SocketIO

class ActorClassesListener: SocketListener, ActorListener {
    let classesPublisher: AnyPublisher<[ActorClassModel]?, Never>
    let socket: SocketIOClient

    private let classesSubject = CurrentValueSubject<[ActorClassModel]?, Never>(nil)

    init(socket: SocketIOClient) {
        classesPublisher = classesSubject.eraseToAnyPublisher()
        self.socket = socket
    }

    func addSocketHandlers() {
        socket.on(SocketEvents.SERVER.ACTOR.SEND.SEND_ACTOR_CLASSES) { data, _ in
            do {
                try self.classesSubject.send(SocketListenerUtility.parseSocketEventDataArray(data))
            } catch let FoundryJSONError.errorMessage(errorMessage) {
                print(errorMessage)
            } catch {
                print(error)
            }
        }
    }

    func requestInitialActorData(actorId: String) {
        socket.emit(SocketEvents.IOS.ACTOR.REQUEST_ACTOR_CLASSES, actorId)
    }

    func isReady() -> Bool {
        true
    }
}

extension SocketEvents.IOS.ACTOR {
    static let REQUEST_ACTOR_CLASSES = "ios:requestActorClasses"
}

extension SocketEvents.SERVER.ACTOR.SEND {
    static let SEND_ACTOR_CLASSES = "server:sendActorClasses"
}

File: character-sheet-companion-server/src/listeners/actor/actorClassesListener.ts
import { Server, Socket } from "socket.io";
import {
  getFoundrySocketFromRoom,
  getIosSocketFromRoom,
} from "../../utilities/SocketUtilities";
import { SOCKET_EVENTS } from "../../constants/events";
import { ActorClasses } from "../../interfaces/actor/ActorClasses";

export const actorClassesListener = (io: Server, socket: Socket): void => {
  const requestActorClasses = (actorId: string) => {
    const foundrySocket = getFoundrySocketFromRoom(io, socket);
    if (foundrySocket) {
      foundrySocket.emit(
        SOCKET_EVENTS.SERVER.ACTOR.REQUEST.REQUEST_ACTOR_CLASSES,
        actorId,
        socket.id
      );
    }
  };

  const receiveActorClasses = (
    actorClasses: ActorClasses,
    iosSocketId: string
  ) => {
    const iosSocket = getIosSocketFromRoom(io, socket, iosSocketId);
    if (iosSocket) {
      iosSocket.emit(
        SOCKET_EVENTS.SERVER.ACTOR.SEND.SEND_ACTOR_CLASSES,
        JSON.stringify(actorClasses)
      );
    }
  };

  socket.on(SOCKET_EVENTS.IOS.ACTOR.REQUEST_ACTOR_CLASSES, requestActorClasses);
  socket.on(
    SOCKET_EVENTS.FOUNDRY.ACTOR.SEND_ACTOR_CLASSES,
    receiveActorClasses
  );
};
